DescriptionGenerator: map work#related_links to Cocina description

jmartin-sul opened the pull request:
## Why was this change made?

closes #341 

## How was this change tested?

unit tests, rails console

## Which documentation and/or configurations were updated?

n/a, hopefully specs cover it

mjgiarlo commented on app/services/description_generator.rb:
```diff
@@ -146,5 +147,18 @@ def contributor(work_form_contributor)
+  def related_resource
+    work.related_links.map do |rel_link|
+      uri_hash = {
+        "uri": rel_link.url
```
I think `"uri":` here should be `uri:` in line with the changes in https://github.com/sul-dlss/happy-heron/pull/407/files#diff-f21f5ecf9be9f56bd3c574ea7cb992626d29a9a41cc2926f9f643eadf3a15c55 Because these keys are not actually strings; they're symbols. Not sure why Sorbet is tricked by this!

jmartin-sul replied:
ah, great!  yeah, i thought the `"string":` construction for hash keys was a bit odd, but i figured i'd stay consistent.   mean to ask but forgot, happy to change.

mjgiarlo replied:
I wonder if some of this could move into a new `RelatedLink#to_h` method, but it's not a big deal. Also, we don't (yet?) do this in other models. This way, each model could take a little more responsibility for its own representation, and the description generator can handle the formatting external to the model? I could go either way on it. Would like to hear what you think, though.

mjgiarlo commented on spec/services/description_generator_spec.rb:
```diff
@@ -34,6 +34,10 @@
         { name: [{ value: contrib3_name }], role: [{ value: 'Contributing author' }], type: 'person' }
+      ],
+      relatedResource: [
+        { identifier: [{ uri: 'http://my.awesome.research.io', displayLabel: 'My Awesome Research' }] },
```
It might be nice to have an example that shows a RelatedLink with a missing title. Worth making factory & spec tweaks for this?

jmartin-sul replied:
good call, will do...

mjgiarlo commented on app/services/description_generator.rb:
```diff
@@ -146,5 +147,18 @@ def contributor(work_form_contributor)
+
+  sig { returns(T::Array[T::Hash[String, T::Array[T::Hash[String, String]]]]) }
+  def related_resource
+    work.related_links.map do |rel_link|
```
I wonder if some of this could move into a new RelatedLink#to_h method, but it's not a big deal. Also, we don't (yet?) do this in other models. This way, each model could take a little more responsibility for its own representation, and the description generator can handle the formatting external to the model? I could go either way on it. Would like to hear what you think, though.

jmartin-sul replied:
i like this idea, though i'd also worry slightly about the `#to_h` representation being very cocina specific... but maybe that's ok?  it does seem like the main use of a model's hash representation would be for translation to cocina representation.  

looking at this example:  https://github.com/sul-dlss-labs/cocina-descriptive-metadata/pull/115/files

i'm not sure that i'd expect such a cocina specific represenation from `#to_h`.

pref for starting that here in this PR, vs doing this one case in a follow on PR, vs updating all of the hash building in this class in one fell swoop in a separate PR?

could also have a `#to_cocina` method instead of using the `#to_h` method.

mjgiarlo replied:
`#to_cocina` WFM too. Opinions, co-TL, @jcoyne ?

ndushay replied:
i'd prefer `#to_cocina` or even `#to_cocina_hash` or whatever, but wondering if this refactoring should get priority over other work in ready column ...

mjgiarlo replied:
Like I said, not a big deal, so not a high priority.

jmartin-sul commented:
turned to draft, will fix the output to match the example here after i get back from lunch: https://github.com/sul-dlss-labs/cocina-descriptive-metadata/pull/115/files

jmartin-sul commented:
ok, i:
* fixed the weird not-string keys in `#related_resource`
* updated the factory and spec to test for missing link title
* fixed the output to match the structure described in this example:  https://github.com/sul-dlss-labs/cocina-descriptive-metadata/blob/master/h2_cocina_mappings/h2_to_cocina_related_resource.txt

i'll start a branch based on this branch to try out the `#to_h`/`#to_cocina` refactoring (happy to roll that work into this PR if i get that done before this is merged).

rebased on latest `main`, ready for another look, i think.

jmartin-sul commented:
d'oh, `sig` is now incorrect for the newly added method, turning to draft and fixing...

ndushay commented on app/services/description_generator.rb:
```diff
@@ -159,5 +160,17 @@ def contributor(work_form_contributor)
     end
   end
+
+  sig { returns(T::Array[T::Hash[String, T::Array[T::Hash[String, String]]]]) }
```
ew.  but we probably have this in other spots.   this is where maybe we could use the references to cocina types?

jmartin-sul replied:
yeah, this looks yucky.  i'm about to push something that's... a little longer, but much clearer about the expected structure (imho), so hopefully that'll be less yuck?

mjgiarlo reviewed (approved):
LGTM, merge when undrafted.

jmartin-sul commented:
@ndushay and @mjgiarlo here's the `sig` i ended up with, which i find clearer, though it is longer:
```ruby
sig do
  returns(T::Array[
    T.any(
      { type: String, access: { url: T::Array[{ value: String }] } },
      { type: String, access: { url: T::Array[{ value: String }] }, title: T::Array[{ value: String }] }
    )
  ])
end
```
